### DISCUSSION TITLE
I managed to make something pretty neat that integrates with Cookie Monster

### DISCUSSION THREAD

**Discussion by GhostGlitch** [Aktanusa/CookieMonster#891] [Show and tell]
I don't know if anyone is interested, but I made a userscript that among other features automatically buys anything that are Green or better thanks to CookieMonsterData. I know it's probably not the best, but I haven't coded in years and I'm pretty proud of it.

https://greasyfork.org/en/scripts/429271-auto-player-for-cookie-monster

**Comment by DanielNoord** [comment]
Good work! I think I saw you reddit post, good to see you have been able to use our data and make something!

Although I wouldn't want to include it in the base mod, I have been working on a sort of API to integrate more mods within the Cookie Monster Framework. I think this mod would fit in nicely although we would then need to rewrite parts of the code.

A good first step (and nice practice exercise for your javascript) you should consider anyway is using the modding API provided by Cookie Clicker. Although it is quite limited, I believe it is good if all mods start to use it!

**Comment by GhostGlitch** [comment]
Oh for sure, it wouldn't fit with the base mod. Just thought you might be interested to see what your mod inspired. I'm excited to see where that Framework idea goes, I would love anything you can do to let my mod leach off of your hard work lol.

> A good first step (and nice practice exercise for your javascript)

This project is literally where all of my javascript knowledge is from. didn't know anything about it until I decided to hack this together.

Also yeah, I know I should use the API. This was just supposed to be a small quick project of small edits of another person's script. I kinda got carried away and there isn't too much of his code left. I should probably just port it to be a true mod. Are there any docs for the CC API? Not entirely sure where to start with that.

**Comment by DanielNoord** [comment]
If you do port your mod to the CC API it shouldn't be too hard to port it to the CM Framework. The Framework is currently quite basic (and doesn't have documentation) but it would be a good test which I would certainly be willing to help with!

Only documentation there is in the base game code. Check `main.js` on line 1413.

**Reply by GhostGlitch** [reply]
Cool, thanks. Even just the line number is something. I'll look into it tomorrow.

**Comment by GhostGlitch** [comment]
Thanks for the help btw. If you care it's at [GhostCookies](https://github.com/GhostGlitch/GhostCookies) now. I know it's a bit of a mess, but it took me 10+ hours to even get it functional without being a UserScript. Hopefully I can actually get around to integrating with the API properly soon.

**Reply by DanielNoord** [reply]
I think it should still work after recent updates to the dev-branch. Let me know if you need any help with transferring to the API!

**Comment by GhostGlitch** [comment]
When do you think I should have it call autoBuy() and autoClick()? Having them as infinite setTimeout's probably isn't the best way to do it, but also I'm not sure I want to sacrifice having them be on a setable delay.

Probably a dumb question, but do I make my whole mod an object for the API? Or do I just create an object up top that calls everything?

How do I add options to the UI? I'm not loving having everything be through the console. Even just knowing where cookie monster does it would be helpful because then I can look at your code and copy how you do it.

Thanks for your help, I kinda threw myself into the project with 0 experience.

**Reply by DanielNoord** [reply]
I think adding to those to the `logic` hook should work fine. Creating the delay could be done by comparing the "delay-time" with the game time. We do something similar [here](https://github.com/CookieMonsterTeam/CookieMonster/blob/dev/src/Cache/CPS/CPS.js) at line 38.

I don't really understand your second question. What you should do is create an object with 3 functions: `init`, `load`, and `save`. Within the `init` function you can then register other functions, for example `autoBuy()` with the `logic` hook.

Adding options to the UI would be best done via the Cookie Monster Framework, although that functionality is not totally finished yet. After you have transferred to the Modding API I will help you incorporate CMF. 
Our current code is in `src/Disp/MenuSections/Settings`. However, I want most of those functions to become part of the CMF and then have CM call them from CMF. You can look at `src/Disp/MenuSections/createMenuInfo.js` for a similar example where CM calls functions from CMF to construct a Info screen menu which is then appended to a CMF-div in the info section.

No problem, I did the same when I started on Cookie Monster and I have learned quite a lot! It is a very nice challenge!
